src/dramatiq_abort/abort_manager.py

```diff
@@ -0,0 +1,142 @@
+import abc
+import contextlib
+import threading
+import time
+from logging import Logger
+from threading import Thread
+from typing import Any, ContextManager, Dict, List, Optional, Tuple
+
+from dramatiq.logging import get_logger
+from dramatiq.middleware.threading import Interrupt, raise_thread_exception
+
+
+def is_gevent_active() -> bool:
+    """Detect if gevent monkey patching is active.
+
+    This function should be removed and replaced with the identically named
+    function imported from dramatiq.middleware.threading as soon as a
+    dependency on a version of dramatiq with this function can be made.
+    """
+    try:
+        from gevent import monkey
+    except ImportError:  # pragma: no cover
+        return False
+    return bool(monkey.saved)
+
+
+class Abort(Interrupt):
+    """Exception used to interrupt worker threads when their worker
+    processes have been signaled to abort.
+    """
+
+
+class AbortManager(abc.ABC):
+    """ABC for raising Abort exceptions in threads.
+
+    :param logger: The logger for abort log lines.
+    :type logger: :class:`logging.Logger`
+    """
+
+    logger: Any
+    lock: ContextManager[Any]
+    abortable_messages: Dict[str, Any]
+    abort_requests: Dict[Any, Tuple[str, float]]
+
+    @abc.abstractmethod
+    def __init__(self, logger: Optional[Logger] = None):
+        self.logger = logger or get_logger(__name__, type(self))
```

Comment:
I think we might as well set `abortable_messages` and `abort_requests` here since its quite core to the base.

`lock` could also be defined as an abstract property, forcing implementer to define it, ie:

```
@abc.abstractmethod
@property
def lock(self) -> ContextManager[Any]: ...
  # CTypesManager impl: return self._lock
```